### ISSUE TITLE
Remove finalized events upon startup

### ISSUE THREAD

**Issue by fredo** [beamer-bridge/beamer#1068]
Upon start up the agent will process all events of the contracts' history. This is unnecessary because the longer the contract lives, the more events are already finalized.
The definition of finalized here is, there is no possibility that another event can be emitted to a certain request or claim id. Typically that is the case when the request deposit is withdrawn and all claims associated to the request are withdrawn. That is also the case when the agent deletes its internal representation of the request and the claims.
In fact, this will happen after sync. All finalized objects will be deleted.

Currently we have seen that with more events, unforeseen errors appear in the agent's state machine. Almost all of the errors happen during start up, where (only one of the reasons) events arrive in different orders than during runtime.

In order to make the start up more efficient and reduce the probability of errors during start up , we can scan the event loop for all finalized events and remove them before processing. It should result in the same outcome, as all finalized request and claim objects will be removed immediately.

## Filtering for finalized events
My suggestion is the following: 

* fetch all events from both chains
* sort target chain events before source chain events
* iterate backwards over the event loop (withdrawn events always the last events in the lifetime of requests and claims)
* mark any claim id or request id which gets withdrawn
* collect all other events with the request or claim id in it
* If requests and all associated claims are completely finalized, remove the events from the event loop
* start with processing events

### When are requests and claims finalized?
A request is finalized if all associated claims are finalized and the deposit is withdrawn

A claim is finalized if all stakes are withdrawn. Note, that with multiple challengers, we can see multiple `ClaimStakeWIthdrawn` events for a given claim id. Unfortunately we dont emit the amount withdrawn with the event. The solution to this is to match all `ClaimMade` events which belong to challengers with `ClaimStakeWithdrawn` events. In this way we can see if all challengers withdrew their stake. If the claimer, is the one who withdrew, then there can only be one `ClaimStakeWithdrawn` event. This can be checked with the original `ClaimMade` event (lastChallenger == address(0) or totalClaimStake == 0).

I know that searching for finalized claims is a bit more complicated, but I think it reduces the probability of unforeseen errors during start up by a lot.

**Comment by fredo** [comment]
I briefly looked at the code and luckily the initial sync is already decoupled from the infinite loop in the EventProcessor. 
This gives us the possibility to do something like this

```
@@ -162,6 +162,8 @@ class EventProcessor:
             self._have_new_events.wait(poll_period)
             self._have_new_events.clear()
 
+        filter_finalized_events(self._events)
+
         while not self._stop:
             if self._have_new_events.wait(poll_period):
                 self._have_new_events.clear()
```

**Comment by fredo** [comment]
> First, fetching all events from "both chains" will become "fetching all events from "all chain", since we are planning to support multiple chain pairs. We may end up waiting needlessly for some random chain's RPC, while we could do some useful work. And the requirement to sort events makes things worse.

We are doing this already and I actually don't know how you would want to prevent that. We are not processing events before the sync is done.

> Second, it is not obvious to me how sorting would even work in case there are multiple source and multiple target chains. We will basically have an any-to-any situation.

That's easy. In order to make it work we need to sort that Request Manager events are at the end of the queue. This is achievable by iterating once over the queue which is of complexity O(n).

> Another issue is that of finalization logic. The way I am understanding the description leads me to imagine two mechanisms for finalization: one during sync, the other during regular operation. The latter would work same as it is working now, based on state changes due to processed events, however, the former cannot operate like that due to events not being processed (meaning no state changes occur). This is not nice because we will have to maintain two separate finalization mechanisms.

I don't understand this part. We are simply discarding events as if they never happened. 

> Yet another issue is that of scalability. With small number of events and objects, this is may be viable, but as the number of events (and chains) increases, the situation worsens.

Why is that so? Given the fact that we actually fetch all events at start up.

> Instead of doing this, I would rather implement the idea we already (briefly) mentioned: representing contract state correctly, with a low-level state machine (let's call this CSR - Contract State Representation, for lack of better naming). With CSR combined with your idea of decoupling the Filler and Challenger state machines, I think we would
>
> solve all the event ordering issues on the CSR level
> have only one finalization mechanism, instead of two
> have no scalability issues, even with multiple chain pairs
> simplify the state machines
> easily & precisely discard finalized requests and claims

I think those ideas are complementary. And can easily done both.

I think some of the points you mentioned are out of scope of this issue and have nothing to do with it.

The way the agent works right now: 

1) fetch all historic events from both chains
2) start processing all events
3) remove all finalized state 

-> Outcome: Same is if those events would have never happened.

As I mentioned, the first point is already implemented in the agent and the `CSR` is complementary to my idea. So the main argument here is scalability. The real difference in doing it or not is how do both approaches scale? I gave it a little bit more thought. Let's assume we compare CSR implemented with CSR implemented + filtering

Let the number of events be N
the number of finalized events F =< N
Number of "live" events L = N - F

A) Filtering

In the case of filtering we iterate over the historic event loop 2 times (Sorting, finding finalized events) (I don't know exactly how removing works, but one could safe the indices so this would mean O(F)). Then for each event which gets processed we do a bunch of Memory state changes and logic to be executed. Let's call this a `Factor M`. We would then iterate over the live events and do a state change accordingly -> M*O(L).

So we end up with: `2*O(N) + O(F) + M*O(L)`

B) No filtering
That's easier. We simply process all historic events. 

`M*O(N)`

In terms of scalability we would have to think about if `M*O(N) > 2*O(N) + O(F) + M*O(L)`. 

In other words: 

`M*O(N) = M*O(L) + M*O(F)`

`M*O(F) > 2*O(N) + O(F)` 

`M*O(F) > 3*O(F) + 2*O(L)`

(No guarantees for correct conversion here) 

I think it's hard to reason about these theoretical terms. What I can say is that we can assume that over time `L` will be more or less constant. So with a growing `N`, `F` grows linearly. The rationale behind it is that life events comes from daily usage and eventually they'll all finalize.
To be fair in this calculation we also have to do some memory storage on finding finalized events (storing events in a dictionary) so there should also be an overhead per finalized event.

With growing F `2*O(L)` can be neglected

so it boils down to `M*O(F) > 3*O(F)`. 

Puh, that was a fun time travel to second semester of CS studies.

**Comment by istankovic** [comment]
> As I mentioned, the first point is already implemented in the agent and the CSR is complementary to my idea. 

Actually, it is not. With CSR, filtering does not make sense. CSR basically does just what the contracts do. I'd be happy to discuss this in a call.

> Let's assume we compare CSR implemented with CSR implemented + filtering

I think we have a different understanding of how CSR would work. Filtering on top of CSR does not make sense -- there is simply nothing to do.

**Comment by fredo** [comment]
Ok. I understand that CSR represents the contract state. So if we have 10 events regarding an object it would modify the state 10 times and ultimately remove it. Am I wrong?

**Comment by istankovic** [comment]
> Ok. I understand that CSR represents the contract state. So if we have 10 events regarding an object it would modify the state 10 times and ultimately remove it. Am I wrong?

No, that's exactly right.

**Comment by fredo** [comment]
Okay, then we would basically do not do this at all. Because we know beforehand that it will be removed again anyways so we can "safe" writing state 10 times. 
Additionally, and most likely we will have to see how we pass events to the state machines of the "roles" like LP and Watcher and if we have to replay state here.

**Comment by fredo** [comment]
The only argument for doing it is potentially a short term relief on potential sync errors because implementing CSR is most likely not be done in the short term.

**Comment by istankovic** [comment]
> I still don't understand why you argue against this with CSR because filtering comes one layer before and I think it potentially speeds up sync even with CSR but all in all it seems to be a nice to have instead of a must have.

CSR, by necessity, must implement finalization logic. Adding another finalization logic and a step before the events are fed to CSR seems completely unnecessary. Consider that you need to fetch all events, sort them (which is `NlogN` at best) and then visit _each_ object just to check whether it refers to one of the already finalized objects. I think we can agree that that is for sure more costly than a simple state change.

**Comment by fredo** [comment]
> > I still don't understand why you argue against this with CSR because filtering comes one layer before and I think it potentially speeds up sync even with CSR but all in all it seems to be a nice to have instead of a must have.
> 
> CSR, by necessity, must implement finalization logic. Adding another finalization logic and a step before the events are fed to CSR seems completely unnecessary. Consider that you need to fetch all events, sort them (which is `NlogN` at best) and then visit _each_ object just to check whether it refers to one of the already finalized objects. I think we can agree that that is for sure more costly than a simple state change.

I think you have not understood what I just wrote above. It's not `nlogn`.

**Comment by istankovic** [comment]
> The only argument for doing it is potentially a short term relief on potential sync errors

Pretty much, though I am quite sceptical about the benefit here.

> because implementing CSR will most likely not be done in the short term.

Well, that would depend on prioritization. Personally, I would not postpone CSR, or something that improves the current situation at least that much, for too long because we will end up complicating the code base further and further, which is not sustainable.

**Comment by istankovic** [comment]
> Not understanding my above described complexity break down is not a basis for a discussion. 

Fully agreed. :+1: 
I will reread the description once more to see what I missed.

However, my point still holds even you throw out the sort operation altogether or make it `O(1)`.

**Comment by fredo** [comment]
I really dont understand why you write `We need to fetch all events`. We have to do that in any case so this can't be part of an evaluation.

Take this as an illustration:
```
for event in events:

    if is_instance(event, RequestManagerEvent):
        request_manager_events.append(event)
    
    if is_instance(event, FillManagerEvent):
        fill_manager_events.append(event)

fill_manager.extend(request_manager)
```

Does not seem to me as nlogn.

**Comment by istankovic** [comment]
> I really dont understand why you write `We need to fetch all events`. We have to do that in any case so this can't be part of an evaluation.
> 
> Take this as an illustration:
> 
> ```
> for event in events:
> 
>     if is_instance(event, RequestManagerEvent):
>         request_manager_events.append(event)
>     
>     if is_instance(event, FillManagerEvent):
>         fill_manager_events.append(event)
> 
> fill_manager.extend(request_manager)
> ```
[...]
> Does not seem to me as `nlogn`.

Eh, if the above is what you meant by "sort", then that explains my confusion. Nevertheless, as I mentioned, even if you ignore the sort operation, you still do work that is not necessary (note: this is all in response to the "CSR + filtering?" topic).

**Comment by istankovic** [comment]
I would like to expand on this a bit:

> With a potential new state machine, filtering can happen on different levels (before CSR, after CSR, etc.)

The CSR is in the best position to do filtering i.e. prevent the unnecessary events reaching the higher (role) level, by doing
the following:

- during sync, events for each object (request or claim) are being stored separately, in order (e.g. `dict[RequestHash, [Event])`)
- as the CSR consumes each event and updates its state, some of the objects will become finalized
- once an object is finalized, the entire sequence of events for that object is dropped
- once the sync is done, the CSR delivers all remaining events (related to still live objects) to the higher level at once
- the higher level objects are now able to correctly reconstruct their start
- event processing continues as usual, with CSR forwarding necessary events as they come
